File: src-tauri/src/commands.rs
mod app_state;
mod helpers;
mod http;
mod platform;
#[cfg(desktop)]
mod runtime;

use serde_json::Value;
use tauri::AppHandle;
#[cfg(desktop)]
use tauri::State;

use crate::models::{
    ActivityPayload, ApiResult, AppStatePayload, ClientCapabilities, ClientConfig,
    DiscordPresenceSnapshot, ExistingReporterConfig, ImportedIntegrationConfig,
    InspirationEntryCreateInput, PlatformSelfTestResult, RealtimeReporterSnapshot,
};
#[cfg(desktop)]
use crate::{discord_presence::DiscordPresenceRuntime, realtime_reporter::ReporterRuntime};

#[tauri::command]
pub fn load_app_state(app: AppHandle) -> Result<AppStatePayload, String> {
    app_state::load_app_state(app)
}

#[tauri::command]
pub fn save_app_state(
    app: AppHandle,
    payload: Value,
) -> Result<ApiResult<AppStatePayload>, String> {
    app_state::save_app_state(app, payload)
}

#[tauri::command]
pub fn restart_app(app: AppHandle) -> Result<(), String> {
    app_state::restart_app(app)
}

#[tauri::command]
pub fn parse_imported_integration_config(
    input: String,
) -> Result<ApiResult<ImportedIntegrationConfig>, String> {
    app_state::parse_imported_integration_config(input)
}

#[tauri::command]
pub fn get_client_capabilities() -> Result<ApiResult<ClientCapabilities>, String> {
    app_state::get_client_capabilities()
}

#[cfg(desktop)]
#[tauri::command]
pub fn hide_to_tray(app: AppHandle) -> Result<(), String> {
    runtime::hide_to_tray(app)
}

#[tauri::command]
pub async fn submit_activity_report(
    config: ClientConfig,
    payload: ActivityPayload,
) -> Result<ApiResult<Value>, String> {
    http::submit_activity_report(config, payload).await
}

#[tauri::command]
pub async fn get_public_activity_feed(config: ClientConfig) -> Result<ApiResult<Value>, String> {
    http::get_public_activity_feed(config).await
}

#[tauri::command]
pub async fn list_inspiration_entries(
    config: ClientConfig,
    limit: Option<u32>,
    offset: Option<u32>,
) -> Result<ApiResult<Value>, String> {
    http::list_inspiration_entries(config, limit, offset).await
}

#[tauri::command]
pub async fn probe_connectivity(config: ClientConfig) -> Result<ApiResult<Value>, String> {
    http::probe_connectivity(config).await
}

#[tauri::command]
pub async fn create_inspiration_entry(
    config: ClientConfig,
    input: InspirationEntryCreateInput,
) -> Result<ApiResult<Value>, String> {
    http::create_inspiration_entry(config, input).await
}

#[tauri::command]
pub async fn upload_inspiration_asset(
    config: ClientConfig,
    image_data_url: String,
) -> Result<ApiResult<Value>, String> {
    http::upload_inspiration_asset(config, image_data_url).await
}

#[cfg(desktop)]
#[tauri::command]
pub fn start_realtime_reporter(
    app: AppHandle,
    reporter: State<'_, ReporterRuntime>,
    config: ClientConfig,
) -> Result<ApiResult<RealtimeReporterSnapshot>, String> {
    runtime::start_realtime_reporter(app, reporter, config)
}

#[cfg(desktop)]
#[tauri::command]
pub fn stop_realtime_reporter(
    reporter: State<'_, ReporterRuntime>,
) -> Result<ApiResult<RealtimeReporterSnapshot>, String> {
    runtime::stop_realtime_reporter(reporter)
}

#[cfg(desktop)]
#[tauri::command]
pub fn get_realtime_reporter_snapshot(
    reporter: State<'_, ReporterRuntime>,
) -> Result<ApiResult<RealtimeReporterSnapshot>, String> {
    runtime::get_realtime_reporter_snapshot(reporter)
}

#[cfg(desktop)]
#[tauri::command]
pub fn start_discord_presence_sync(
    app: AppHandle,
    discord_presence_runtime: State<'_, DiscordPresenceRuntime>,
    config: ClientConfig,
) -> Result<ApiResult<DiscordPresenceSnapshot>, String> {
    runtime::start_discord_presence_sync(app, discord_presence_runtime, config)
}

#[cfg(desktop)]
#[tauri::command]
pub fn stop_discord_presence_sync(
    discord_presence_runtime: State<'_, DiscordPresenceRuntime>,
) -> Result<ApiResult<DiscordPresenceSnapshot>, String> {
    runtime::stop_discord_presence_sync(discord_presence_runtime)
}

#[cfg(desktop)]
#[tauri::command]
pub fn get_discord_presence_snapshot(
    discord_presence_runtime: State<'_, DiscordPresenceRuntime>,
) -> Result<ApiResult<DiscordPresenceSnapshot>, String> {
    runtime::get_discord_presence_snapshot(discord_presence_runtime)
}

#[tauri::command]
pub async fn run_platform_self_test() -> Result<ApiResult<PlatformSelfTestResult>, String> {
    platform::run_platform_self_test().await
}

#[tauri::command]
pub fn request_accessibility_permission() -> Result<ApiResult<bool>, String> {
    platform::request_accessibility_permission()
}

#[tauri::command]
pub fn discover_existing_reporter_config(
    app: AppHandle,
) -> Result<ApiResult<ExistingReporterConfig>, String> {
    app_state::discover_existing_reporter_config(app)
}

File: src/lib/api/commands.ts
import { invoke } from "@tauri-apps/api/core";

import { translate } from "../../i18n";
import { resolveApiErrorMessage } from "../localizedText";
import type {
  ActivityFeedData,
  ActivityPayload,
  ApiResult,
  ClientCapabilities,
  ClientConfig,
  DiscordPresenceSnapshot,
  ExistingReporterConfig,
  ImportedIntegrationConfig,
  InspirationAssetUploadResult,
  InspirationEntry,
  InspirationEntryCreateInput,
  InspirationEntryListResponse,
  PlatformSelfTestResult,
  RealtimeReporterSnapshot,
} from "../../types";
import { invokeApi } from "./invoke";

export async function getClientCapabilities(): Promise<ApiResult<ClientCapabilities>> {
  return invokeApi("get_client_capabilities");
}

export async function submitActivityReport(
  config: ClientConfig,
  payload: ActivityPayload,
): Promise<ApiResult<Record<string, unknown>>> {
  return invokeApi("submit_activity_report", { config, payload });
}

export async function getPublicActivityFeed(
  config: ClientConfig,
): Promise<ApiResult<ActivityFeedData>> {
  return invokeApi("get_public_activity_feed", { config });
}

export async function listInspirationEntries(
  config: ClientConfig,
  options?: { limit?: number; offset?: number },
): Promise<ApiResult<InspirationEntryListResponse>> {
  return invokeApi("list_inspiration_entries", {
    config,
    limit: options?.limit,
    offset: options?.offset,
  });
}

export async function probeConnectivity(
  config: ClientConfig,
): Promise<ApiResult<Record<string, unknown>>> {
  return invokeApi("probe_connectivity", { config });
}

export async function createInspirationEntry(
  config: ClientConfig,
  input: InspirationEntryCreateInput,
): Promise<ApiResult<InspirationEntry>> {
  return invokeApi("create_inspiration_entry", { config, input });
}

export async function uploadInspirationAsset(
  config: ClientConfig,
  imageDataUrl: string,
): Promise<ApiResult<InspirationAssetUploadResult>> {
  return invokeApi("upload_inspiration_asset", { config, imageDataUrl });
}

export async function parseImportedIntegrationConfig(
  input: string,
): Promise<ImportedIntegrationConfig> {
  const result = await invokeApi<ImportedIntegrationConfig>(
    "parse_imported_integration_config",
    { input },
  );

  if (!result.success || !result.data) {
    throw new Error(
      resolveApiErrorMessage(
        result.error,
        translate,
        translate("connectionPanel.notify.importFailedDetail"),
      ),
    );
  }

  return result.data;
}

export async function hideToTray(): Promise<void> {
  await invoke("hide_to_tray");
}

export async function startRealtimeReporter(
  config: ClientConfig,
): Promise<ApiResult<RealtimeReporterSnapshot>> {
  return invokeApi("start_realtime_reporter", { config });
}

export async function stopRealtimeReporter(): Promise<ApiResult<RealtimeReporterSnapshot>> {
  return invokeApi("stop_realtime_reporter");
}

export async function getRealtimeReporterSnapshot(): Promise<ApiResult<RealtimeReporterSnapshot>> {
  return invokeApi("get_realtime_reporter_snapshot");
}

export async function startDiscordPresenceSync(
  config: ClientConfig,
): Promise<ApiResult<DiscordPresenceSnapshot>> {
  return invokeApi("start_discord_presence_sync", { config });
}

export async function stopDiscordPresenceSync(): Promise<ApiResult<DiscordPresenceSnapshot>> {
  return invokeApi("stop_discord_presence_sync");
}

export async function getDiscordPresenceSnapshot(): Promise<ApiResult<DiscordPresenceSnapshot>> {
  return invokeApi("get_discord_presence_snapshot");
}

export async function runPlatformSelfTest(): Promise<ApiResult<PlatformSelfTestResult>> {
  return invokeApi("run_platform_self_test");
}

export async function requestAccessibilityPermission(): Promise<ApiResult<boolean>> {
  return invokeApi("request_accessibility_permission");
}

export async function discoverExistingReporterConfig(): Promise<ApiResult<ExistingReporterConfig>> {
  return invokeApi("discover_existing_reporter_config");
}

export async function restartApp(): Promise<void> {
  await invoke("restart_app");
}
